airbyte-commons-server/src/main/java/io/airbyte/commons/server/handlers/OAuthHandler.java

```diff
@@ -361,4 +361,51 @@ JsonNode getOAuthInputConfiguration(final JsonNode hydratedSourceConnectionConfi
     return Jsons.jsonNode(result);
   }
 
+  /**
+   * Given an OAuth response, writes a secret and returns the secret Coordinate in the appropriate format.
+   * <p>
+   * Unlike our regular source creation flow, the OAuth credentials created and stored this way will be stored
+   * in a singular secret as a string. When these secrets are used, the user will be expected to use the
+   * specification to rehydrate the connection configuration with the secret values prior to saving a
+   * source/destination.
+   * <p>
+   * The singular secret was chosen to optimize UX for public API consumers (passing them one secret
+   * to keep track of > passing them a set of secrets).
+   * <p>
+   * See https://github.com/airbytehq/airbyte/pull/22151#discussion_r1104856648 for full discussion.
+   */
+  public Map<String, Object> writeOAuthResponseSecret(final UUID workspaceId, final Map<String, Object> payload) {
+
+    try {
+      final String payloadString = Jackson.getObjectMapper().writeValueAsString(payload);
+      final SecretCoordinate secretCoordinate = secretsRepositoryWriter.storeSecret(generateOAuthSecretCoordinate(workspaceId), payloadString);
+      return formatOAuthSecretResponse(secretCoordinate);
+    } catch (final JsonProcessingException e) {
+      throw new RuntimeException("Json object could not be written to string.", e);
+    }
+  }
+
+  public Map<String, Object> formatOAuthSecretResponse(final SecretCoordinate secretCoordinate){
+    return Map.of("secretId", secretCoordinate.getFullCoordinate());
+  }
+
+  /**
+   * Generate our OAuthSecretCoordinate for storage For now we're treating these as always v1.
+   */
+  private SecretCoordinate generateOAuthSecretCoordinate(final UUID workspaceId) {
+    final String coordinateBase = SecretsHelpers.getCoordinatorBase("airbyte_oauth_workspace_", workspaceId, UUID::randomUUID);
+    return new SecretCoordinate(coordinateBase, 1);
+  }
+
+  public Map<String, Object> completeSourceOAuthHandleReturnSecret(final CompleteSourceOauthRequest completeSourceOauthRequest)
```

Comment:
I see we are only doing this for the source - what about the destination? will we do that in a follow up PR? Or do we not need this for the destination for now?

Reply:
nit: I would move this method to be before the `completeSourceOAuth` method on line 172. In general, I try to place the methods so a reader scrolling from top to bottom will encounter methods in order of their importance as they go about reading the class.

Reply:
Don't need this for destination for now (at least for PbA use case) but PE is planning to add it for destinations in the future.

Appreciate the comment, will move the method and will keep that in mind as it's a good rule to follow.